docs/design/2021-08-18-charsets.md

```diff
@@ -0,0 +1,270 @@
+    * [Benchmark Tests](#benchmark-tests)
+* [Investigation & Alternatives](#investigation--alternatives)
+* [Development Plan](#development-plan)
+
+# Introduction
+
+Currently, TiDB only supports ascii, binary, latin1, utf-8, and utf8mb4 character sets. This proposal aims to add a framework for character sets to TiDB to facilitate the support of other characters. This design takes the gbk character set as an example.
+
+# Motivation or Background
+
+Character: Various words and symbols are collectively referred to as characters. A character can be a Chinese character, an English letter, an Arabic numeral, a punctuation mark, a graphic symbol or a control symbol, etc.
+
+Charset: A collection of multiple characters is called a charset. Common character set names: ASCII character set, Unicode character set, GBK character set, etc.
+
+Character Encoding: Character encoding can be considered as a mapping rule. According to this mapping rule, a character can be mapped into other forms of data for storage and transmission in the computer. Each character set has corresponding character encoding rules, and the commonly used character set encoding rules include UTF-8 encoding, GBK encoding and so on.
+
+MySQL supports many character sets, such as `utf8mb4`, `gbk`, `gb18030`, etc. However, currently TiDB only supports 5 character sets, and there are still some problems with related character sets ([charset: incorrect encoding for `latin1` character set](https://github.com/pingcap/tidb/issues/18955)，[extended latin character set not support](https://github.com/pingcap/tidb/issues/3888)). Now it is difficult for TiDB to add a new character set. This proposal describes the implementation of the character set framework. In many traditional industries in China, `gbk`/`gb18030` is used, so this proposal uses gbk character set support as an example.
+
+## Characteristics
+
+Support gbk character set related functions:
+- Supports all code points defined by the gbk standard. Supports all code points defined by the gbk standard.
+- Support conversion with other character sets.
+  - Support data automatically converted according to the corresponding character set.
+  - Supports the use of `CAST`, `CONVERT` and other functions to convert gbk characters.
+- Support SQL statements, such as: `SET CHARACTER SET GBK`, `SET NAMES GBK`, `SHOW CHARSET`, etc.
+- It supports the comparison between strings in the gbk character set, as well as the comparison between gbk strings and strings in other character sets.
+- If the characters passed in are illegal, an error or warning is returned. If an illegal character is used in `CONVERT`, an error is returned. Otherwise, a warning will be returned.
+- Compatible with other components, including data import, migration, and so on.
+
+### Supported Collations
+
+- gbk_bin：A binary collation.
+- gbk_chinese_ci：The default collation, which supports Pinyin.
+- gbk_utf8mb4_bin: A binary collation encoded by utf8mb4.
+
+# Detailed Design
```

Comment:
the detailed design should distinguish between gbk-specific stuff and generic charset (e.g. gb18030 or sjis) stuff.

Reply:
it is not clear from the document whether some design point is specific to GBK (esp. the upgrade and downgrade ones). if we introduce more than GBK in the same version (e.g. GB18030 as well) it is not clear what could be applicable to the other charsets.

Reply:
https://github.com/pingcap/tidb/pull/27325/files#:~:text=This%20design%20takes%20the%20gbk%20character%20set%20as%20an%20example is explained that this design is based on GBK as an example, and in fact, GBK was used in the front, and the use of non-UTF-8 here is not very meaningful

Reply:
The proposal title is "**Support Charset Framework in TiDB**", not "**Support GBK in TiDB**". It is not good if we cannot distinguish between what part of the proposal is GBK-only and what part is applicable to the entire "charset framework".